cli/doc/man/man1/virgilpub2cert.1
.Dd April 22, 2015
.Dt PUB2CERT 1
.Os Darwin
.Sh NAME
.Nm pub2cert
.Nd create certificates
.Sh SYNOPSIS
.Nm
.Op Fl Fl in Ar file
.Op Fl Fl out Ar file
.Op Fl Fl format Ar arg
.Op Fl Fl account-id Ar arg
.Fl Fl certificate-id Ar arg
.Sh DESCRIPTION
Create certificate from the public key and identifiers.
.Sh OPTIONS
.Bl -tag -width "--"
.It Fl i, Fl Fl in Ar file
Public key. If omitted \fBstdin\fP is used.
.It Fl o, Fl Fl out Ar file
Certificate. If omitted \fBstdout\fP is used.
.It Fl f, Fl Fl format Ar arg
Output format (\fBjson\fP | \fBder\fP - default).
.It Fl a, Fl Fl account-id Ar arg
Account identifier.
.It Fl c, Fl Fl certificate-id Ar arg
Certificate identifier.
.El
.Sh SEE ALSO
.Xr virgil 1 ,
.Xr virgilcertinfo 1


ChangeLog
virgil ChangeLog (Sorted per date)

= Version 0.8.0 released 2015-04-16

Bugfix:
  * Lib: Store algorithm parameters in ASN.1 structure when marshal cipher with elliptic-curve key.
  * Programs: Add omitted parameter 'certificate id'.

Features:
  * AS3: Add class VirgilBase64.
  * Ruby. Add wrapper for the ruby.
  * Python. Add wrapper for the python.
  * Emscripten: Add emscripten(asm.js) wrapper.

Improvements:
  * Lib: Encapsulate fields of the class VirgilCipherBase.
  * Lib: Hide usage of classes: VirgilAsn1Reader and VirgilAsn1Writer.
  * CMake: Remove redundant installation of library headers.
  * CMake: Add optional flag LOW_LEVEL_API - default value is OFF.
  * CMake: Remove unused variable - SIMULATOR.

Changes:
  * AS3: Move classes 'VirgilCustomParams' and 'VirgilContentInfo' to the package 'com.virgilsecurity.crypto'.
  * Lib: Add member type 'value_type' to the VirgilIdProvider.
  * Lib: Move version configuration from header file to the implementation file.
  * CMake: Change library installation policy, now target folder depends on the machine architecture.

= Version 0.7.0 released 2015-03-31
Bugfix:
  * CMake: Fix CMAKE_OSX_ARCHITECTURES variable default definition.

Features:
  * C#: Add mapping C++ VirgilByteArray to the C# byte[].
  * C#: Add runtime selection of the library architecture: x86 or x86_64.
  * AS3: Add wrapper for class VirgilHash.

Improvements:
  * C#: Set assembly version to the library current version.
  * AS3: Move utility headers to the appropriate place.
  * SWIG: Simplify including type mappings for VirgilByteArray.
  * SWIG: Move language specific typemaps to separate files.

Changes:
  * All: Rename method VirgilChunkCipher::finalize() to
         VirgilChunkCipher::finish() due to name confilcts on some platforms.
  * SWIG C#: Make first letter of all functions and properties uppercase.
  * CMake C#. Rename .NET assembly of the shared library from 'virgil.assembly' to 'Virgil.Net'
  * CMake: Selection of install directory for C# now depends on variables:
           CMAKE_SYSTEM_NAME, CMAKE_OSX_ARCHITECTURES, COMPILER_ARCH_32,
           COMPILER_ARCH_64.
  * SWIG: Ignore 'operator=' and 'operator=='.

= Version 0.6.0 released 2015-03-23
Bugfix:
  * CMake: Fix build options for jsoncpp library.
  * CMake: Define max java heap size for AS3 compilation process to prevent
           build fail.
  * AS3: Add try-catch section to the 'VirgilChunkCipher::startEncryption()'
         method wrapper.

Improvements:
  * CMake: Add version to the AS3 library file name.
  * CMake: Display full AS3 library name (including version) when build.
  * CMake: Prevent copy C++ headers to the AS3 installation folder.
  * CMake: Make AS3 library build only if needed.
  * CMake: Replace variable AS3_MODULE_FILE_NAME with synonym WRAPPER_SWC.
  * AS3: Change gradle configuration file to work with gradle version 2.3
       and gradlefx version 1.1.0.
  * CMake: Use UPDATE command instead of PATCH command in the
           ExternalProject_Add.
  * Lib: Add SHA-384 algorithm.
  * CMake: Add AS3 wrapper dependency on the library 'virgil'.

Changes:
  * AS3 Demo: Change demo app to be conformed with library version 0.5.0.
  * AS3 Demo: Embed content info to the encrypted file.
  * PHP: Remove 'iconv' and 'mb_detect_encoding' dependencies from unit tests.
  * All: Switch from 'polarssl' main branch to the 'mbed TLS' fork made by
         Virgil Security.
  * CMake: Refactor 'mbed TLS' patch. Now its configure files: 'config.h'
           and 'check_config.h' only.
  * Lib: PBE: use external functions.
  * Lib: Change SHA-256 to SHA-384 as default signer hash algorithm.
  * Lib: Rename VirgilByteArray utility functions to short versions.
  * Lib: Function 'getContentInfo()' become available immediately after
         function 'startEncryption()' call in the class VirgilChunkCipher.
  * AS3: Build AS3 unit tests in the strict mode.

= Version 0.5.0 released 2015-03-12
Features:
  * AS3: Implement Action Script 3 wrapper for virgil library version 0.4.0.

Improvements:
  * AS3: Speed and size optimization by using flascc Link Time Optimizer (LTO).
  * AS3: Make all tests configurable.
  * AS3: Configure test VirgilVersionTest with CMAKE variables.

Changes:
  * CMake: Apply platform specific compiler options only for native platforms
           build.
  * Lib: Changed marshalling / demarshalling process for tickets.
         Now 'class_name' field is handled by the base class.
  * Lib: Rename VirgilCipherBase::customParameters() to
         VirgilCipherBase::customParams().
  * Lib: Renaming VirgilVersion class methods.
  * Lib: Remove macroses: VIRGIL_BYTE_ARRAY_FROM_C_STRING,
         VIRGIL_BYTE_ARRAY_FROM_STD_STRING, VIRGIL_BYTE_ARRAY_TO_STD_STRING,
         VIRGIL_BYTE_ARRAY_ZEROIZE. Use inline functions analogue instead.
  * Lib: Remove 'jsoncpp' library dependency from the public API.


cli/src/virgilpub2cert.cxx
/**
 * Copyright (C) 2014 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <iostream>
#include <fstream>
#include <algorithm>
#include <iterator>
#include <string>
#include <stdexcept>

#include <virgil/VirgilByteArray.h>
using virgil::VirgilByteArray;

#include <virgil/VirgilException.h>
using virgil::VirgilException;

#include <virgil/service/data/VirgilCertificate.h>
using virgil::service::data::VirgilCertificate;

#include <tclap/CmdLine.h>

#include "utils.h"

#ifdef SPLIT_CLI
    #define MAIN main
#else
    #define MAIN pub2cert_main
#endif

int MAIN(int argc, char **argv) {
    try {
        // Parse arguments.
        TCLAP::CmdLine cmd("Create certificate from the public key and identifiers.", ' ', virgil::cli::version());

        TCLAP::ValueArg<std::string> inArg("i", "in", "Public key. If omitted stdin is used.",
                false, "", "file");

        TCLAP::ValueArg<std::string> outArg("o", "out", "Certificate. If omitted stdout is used.",
                false, "", "file");

        TCLAP::ValueArg<std::string> formatArg("f", "format", "Output format: json | der (default).",
                false, "der", "arg");

        TCLAP::ValueArg<std::string> accountIdArg("a", "account-id", "Account identifier.",
                false, "", "arg");

        TCLAP::ValueArg<std::string> certificateIdArg("c", "certificate-id", "Certificate identifier.",
                true, "", "arg");

        cmd.add(certificateIdArg);
        cmd.add(accountIdArg);
        cmd.add(formatArg);
        cmd.add(outArg);
        cmd.add(inArg);

        cmd.parse(argc, argv);

        // Prepare input.
        std::istream *inStream = &std::cin;
        std::ifstream inFile(inArg.getValue().c_str(), std::ios::in | std::ios::binary);
        if (inFile.good()) {
            inStream = &inFile;
        } else if (!inArg.getValue().empty()) {
            throw std::invalid_argument(std::string("can not read file: " + inArg.getValue()));
        }

        // Prepare output.
        std::ostream *outStream = &std::cout;
        std::ofstream outFile(outArg.getValue().c_str(), std::ios::out | std::ios::binary);
        if (outFile.good()) {
            outStream = &outFile;
        } else if (!outArg.getValue().empty()) {
            throw std::invalid_argument(std::string("can not write file: " + outArg.getValue()));
        }

        // Read public key.
        VirgilByteArray publicKey;
        std::copy(std::istreambuf_iterator<char>(*inStream), std::istreambuf_iterator<char>(),
                std::back_inserter(publicKey));

        // Create certificate
        VirgilCertificate certificate(publicKey);
        certificate.id().setAccountId(virgil::str2bytes(accountIdArg.getValue()));
        certificate.id().setCertificateId(virgil::str2bytes(certificateIdArg.getValue()));

        // Marshal certificate
        VirgilByteArray certificateData;
        if (formatArg.getValue() == "der") {
            certificateData = certificate.toAsn1();
        } else if (formatArg.getValue() == "json") {
            certificateData = certificate.toJson();
        } else {
            throw std::invalid_argument(std::string("unknown --format: ") + formatArg.getValue());
        }

        // Output marshalled certificate
        std::copy(certificateData.begin(), certificateData.end(), std::ostreambuf_iterator<char>(*outStream));

        return EXIT_SUCCESS;
    } catch (TCLAP::ArgException& exception) {
        std::cerr << "Error: " << exception.error() << " for arg " << exception.argId() << std::endl;
        return EXIT_FAILURE;
    } catch (std::exception& exception) {
        std::cerr << "Error: " << exception.what() << std::endl;
        return EXIT_FAILURE;
    }
}
